framework/src/testing/packet.rs

```diff
@@ -0,0 +1,253 @@
+
+/// Enumeration of settable packet fields
+#[allow(non_camel_case_types)]
+#[derive(Debug, Eq, Hash, PartialEq)]
+pub enum field {
+    // ethernet
+    eth_src,
+    eth_dst,
+    // ipv6
+    ipv6_src,
+    ipv6_dst,
+    ipv6_dscp,
+    ipv6_ecn,
+    ipv6_flow_label,
+    ipv6_hop_limit,
+    // tcp
+    tcp_src_port,
+    tcp_dst_port,
+    tcp_seq_no,
+    tcp_ack_no,
+    tcp_window,
+    tcp_urgent_pointer,
+    tcp_ns,
+    tcp_cwr,
+    tcp_ece,
+    tcp_urg,
+    tcp_ack,
+    tcp_psh,
+    tcp_rst,
+    tcp_syn,
+    tcp_fin,
+}
+
+/// `HashMap` of packet fields to their corresponding proptest strategy
+///
+/// Use the `fieldmap!` macro to define fields with their default values.
+/// The fields with defaults are fixed to that value. All other fields
+/// will use the `Any` strategy to generate random values.
+///
+/// # Example
+///
+/// ```
+/// let map = fieldmap! {
+///     field::ipv6_dst_port => 80,
+///     field::tcp_syn => true,
+/// }
+///
+/// When converting default value to proptest strategy, if the type of the
+/// value does not match the field type, the conversion will `panic`.
+/// ```
+pub struct StrategyMap(HashMap<field, Box<Any>>);
+
+impl StrategyMap {
+    pub fn new(inner: HashMap<field, Box<Any>>) -> Self {
+        StrategyMap(inner)
+    }
+
+    fn get<T: Arbitrary + Clone + 'static>(&self, key: &field) -> impl Strategy<Value = T> {
+        if let Some(ref v) = self.0.get(key) {
+            let v = v
+                .downcast_ref::<T>()
+                .unwrap_or_else(|| panic!("value doesn't match type for field '{:?}'", key));
+            Just(v.clone()).boxed()
+        } else {
+            any::<T>().boxed()
+        }
+    }
+
+    fn bool(&self, key: &field) -> impl Strategy<Value = bool> {
+        self.get::<bool>(key)
+    }
+
+    fn u8(&self, key: &field) -> impl Strategy<Value = u8> {
+        self.get::<u8>(key)
+    }
+
+    fn u16(&self, key: &field) -> impl Strategy<Value = u16> {
+        self.get::<u16>(key)
+    }
+
+    fn u32(&self, key: &field) -> impl Strategy<Value = u32> {
+        self.get::<u32>(key)
+    }
+
+    fn mac_addr(&self, key: &field) -> impl Strategy<Value = MacAddr> {
+        self.get::<MacAddr>(key)
+    }
+
+    fn ipv6_addr(&self, key: &field) -> impl Strategy<Value = Ipv6Addr> {
+        self.get::<Ipv6Addr>(key)
+    }
+}
+
+#[macro_export]
+macro_rules! fieldmap {
+    ($($key:expr => $value:expr),*) => {
+        {
+            let mut hashmap = ::std::collections::HashMap::<$crate::testing::field, Box<::std::any::Any>>::new();
+            $(
+                hashmap.insert($key, Box::new($value));
+            )*
+            $crate::testing::StrategyMap::new(hashmap)
+        }
+    };
+}
+
+fn ethernet(ether_type: EtherType, map: &StrategyMap) -> impl Strategy<Value = Ethernet> {
+    (map.mac_addr(&field::eth_src), map.mac_addr(&field::eth_dst)).prop_map(move |(src, dst)| {
+        let packet = RawPacket::new().unwrap();
```

Comment:
what's the reasoning for having an optional on a constructor? wouldnt this always return `Some` and not `None`?

Reply:
a new raw packet will allocate a new dpdk mbuf under the cover which can actually fail if your system is out of memory, out of preallocated mbuf, or possibly for other reasons. so the constructor can in fact fail. the return type on new is `Result<RawPacket, Error>`